backends/occa/ceed-occa-restrict.c

```diff
@@ -35,277 +24,161 @@ static inline size_t bytes(const CeedElemRestriction res) {
-  // occaMemoryFree(*impl->device);
-  // free device object
-  //ierr = CeedFree(&impl->device); CeedChk(ierr);
-  // free our CeedElemRestriction_Occa struct
-  //ierr = CeedFree(&res->data); CeedChk(ierr);
-  ierr = CeedFree(&impl->indices_allocated); CeedChk(ierr);
-  ierr = CeedFree(&r->data); CeedChk(ierr);
+  occaMemoryFree(*data->d_indices);
+  occaKernelFree(data->kRestrict[0]);
+  occaKernelFree(data->kRestrict[1]);
+  occaKernelFree(data->kRestrict[2]);
+  occaKernelFree(data->kRestrict[3]);
+  occaKernelFree(data->kRestrict[4]);
+  occaKernelFree(data->kRestrict[5]);
+  ierr = CeedFree(&data->h_indices); CeedChk(ierr);
+  ierr = CeedFree(&data->d_indices); CeedChk(ierr);
+  ierr = CeedFree(&data); CeedChk(ierr);
   return 0;
 }
 
+// *****************************************************************************
+// * CeedElemRestrictionCreate_Occa
 // *****************************************************************************
 int CeedElemRestrictionCreate_Occa(const CeedElemRestriction r,
                                    const CeedMemType mtype,
                                    const CeedCopyMode cmode,
                                    const CeedInt *indices) {
-  const Ceed_Occa *ceed_data=r->ceed->data;
-  CeedElemRestriction_Occa *impl;
+  CeedDebug("\033[35m[CeedElemRestriction][Create]");
   int ierr;
-
+  CeedElemRestriction_Occa *data;
+  const Ceed_Occa *occa=r->ceed->data;
+  // ***************************************************************************
   if (mtype != CEED_MEM_HOST)
     return CeedError(r->ceed, 1, "Only MemType = HOST supported");
-  // Allocating impl & device **************************************************
+  // Set the functions *********************************************************
+  r->Apply = CeedElemRestrictionApply_Occa;
+  r->Destroy = CeedElemRestrictionDestroy_Occa;
+  // Allocating occa & device **************************************************
   CeedDebug("\033[35m[CeedElemRestriction][Create] Allocating");
-  ierr = CeedCalloc(1,&impl); CeedChk(ierr);
-  ierr = CeedCalloc(1,&impl->device); CeedChk(ierr);
-  *impl->device = occaDeviceMalloc(ceed_data->device, bytes(r), NULL, NO_PROPS);
-  r->data = impl;
+  ierr = CeedCalloc(1,&data); CeedChk(ierr);
+  ierr = CeedCalloc(1,&data->d_indices); CeedChk(ierr);
+  *data->d_indices = occaDeviceMalloc(*occa->device, bytes(r), NULL, NO_PROPS);
+  r->data = data;
+  assert(indices);
   // ***************************************************************************
   switch (cmode) {
+  // Will copy the values and not store the passed pointer
   case CEED_COPY_VALUES:
     CeedDebug("\t\033[35m[CeedElemRestriction][Create] CEED_COPY_VALUES");
-    ierr = CeedMalloc(r->nelem*r->elemsize, &impl->indices_allocated);
-    CeedChk(ierr);
-    memcpy(impl->indices_allocated, indices, bytes(r));
-    impl->indices = impl->indices_allocated;
+    ierr = CeedMalloc(r->nelem*r->elemsize, &data->h_indices); CeedChk(ierr);
+    assert(indices);
```

Comment:
How does `indices` change from your assertion above to here? Also, why do we ever use `assert`? It's terrible library practice and the reason we have an error handling mechanism. See R3 of https://figshare.com/articles/xSDK_Community_Package_Policies/4495136